# STA130 HW07 Homework

**Name:** Varak Tanashian

**Date:** November 14th, 2024

**Assignment:** #7

---

## Course Information
- **Course:** STA130 - An Introduction to Statistical Reasoning and Data Science
- **Instructor:** Prof. Scott Schwartz 
- **Term:** Fall 2024

---

*This notebook contains my solutions and work for the Week 07 Homework assignment. The required Chatbot logs are included as well*

##### ChatGPT Summary: Found at the bottom of this notebook  
##### Chat Link 
- https://chatgpt.com/share/67369183-eff4-800a-a32c-01ce5e6ccb9b

### Question 1


1. **Simple Linear Regression vs. Multiple Linear Regression:**
   - Simple Linear Regression models the relationship between a single independent variable (X) and a dependent variable (Y).
   - Multiple Linear Regression models the relationship between multiple independent variables (X<sub>1</sub>, X<sub>2</sub>, ..., X<sub>n</sub>) and a dependent variable (Y).
   - The benefit of Multiple Linear Regression is that it allows you to analyze the effect of each independent variable on the dependent variable, while controlling for the other variables. This provides a more comprehensive understanding of the relationships in the data.

2. **Continuous Variables vs. Indicator Variables in Simple Linear Regression:**
   - Continuous variables are measurable quantities that can take on any value within a range (e.g., height, weight, temperature).
   - Indicator variables (or dummy variables) are binary variables that take on values of 0 or 1 to represent categorical data (e.g., gender, having a credit card).
   - In Simple Linear Regression, continuous variables are used directly as the independent variable (`X`), while indicator variables are used to represent categorical data.

3. **Behavior Change with Indicator Variables in Multiple Linear Regression:**
   - When an indicator variable is introduced alongside a continuous variable in a Multiple Linear Regression model, the model assumes a different linear relationship for each category represented by the indicator variable.
   - The linear form for Simple Linear Regression is <br><br>
   **Y = β<sub>0</sub> + β<sub>1</sub>X + ε**<br><br>
   while for Multiple Linear Regression with an indicator variable it becomes <br><br>
   **Y = β<sub>0</sub> + β<sub>1</sub>X + β<sub>2</sub>I + ε**<br><br>
   where `I` is the indicator variable.

4. **Effect of Interactions in Multiple Linear Regression:**
   - Introducing an interaction between a continuous variable and an indicator variable in a Multiple Linear Regression model allows the effect of the continuous variable to vary depending on the category represented by the indicator variable.
   - The linear form for this case would be <br><br>
   **Y = β<sub>0</sub> + β<sub>1</sub>X + β<sub>2</sub>I + β<sub>3</sub>(X*I) + ε**<br><br>
   Where `X*I` represents the interaction term.

5. **Multiple Linear Regression with Categorical Variables:**
   - When a Multiple Linear Regression model includes only indicator variables derived from a non-binary categorical variable, the model assumes a different linear relationship for each category.
   - The linear form for this case would be <br><br>
   **Y = β<sub>0</sub> + β<sub>1</sub>I<sub>1</sub> + β<sub>2</sub>I<sub>2</sub> + ... + β<sub>n</sub>I<sub>n</sub> + ε**
   where **I<sub>1</sub>, I<sub>2</sub>, ..., I<sub>n</sub>**<br><br> 
   are the binary indicator variables representing the categories.<br><br>
   - The necessary binary variable encodings for the categorical variable are typically done using techniques like one-hot encoding or dummy coding.

### Question 2

1. **Variables:**
   - Outcome: Sales
   - Predictors: Advertisement Spending (continuous), Product Quality (continuous)

2. **Interactions:**
   The effect of Advertisement Spending on Sales may depend on Product Quality, and vice versa, suggesting a potential interaction.

3. **Linear Forms:**
   - Without Interaction: `Sales = β0 + β1(Ad Spending) + β2(Product Quality) + ε`
   - With Interaction: `Sales = β0 + β1(Ad Spending) + β2(Product Quality) + β3(Ad Spending * Product Quality) + ε`

4. **Predictions:**
   - Without Interaction: Plug in Ad Spending and Product Quality values.
   - With Interaction: The effect of one predictor depends on the other.

5. **Categorical Predictors:**
   - Binary Ad Budget (High/Low) instead of continuous Ad Spending.
   - Linear Forms similar, but Ad Budget replaces Ad Spending.

### Question 3

In [2]:
import pandas as pd
import statsmodels.formula.api as smf

# Step 1: Load the dataset
url = "https://raw.githubusercontent.com/pointOfive/stat130chat130/main/CP/CSCS_data_anon.csv"
df = pd.read_csv(url)

# Step 2: Preprocessing the data
# Let's assume we are interested in a binary outcome, e.g., "Feel_connected_yn" (yes/no for feeling socially connected)
# We also assume it is coded as 1 (Yes) and 0 (No). If not, we should convert it.

# Ensuring binary outcome (already provided in binary form in many survey datasets):
df['Feel_connected_binary'] = (df['Feel_connected_yn'] == 'Yes').astype(int)

# Create some example binary and continuous variables for the model:
# Let's choose 'Age' as a continuous predictor and create a binary indicator for 'Gender' (e.g., Female = 1, Male = 0)
df['Female'] = (df['Gender'] == 'Female').astype(int)

# We may also include 'Education_level' as a categorical variable:
# 'C(Education_level)' will be treated as categorical in the formula notation.

# Step 3: Define and fit the logistic regression model using smf (following the additive form example)
# Additive model specification
formula_additive = 'Feel_connected_binary ~ Age + Female + C(Education_level)'
log_reg_additive = smf.logit(formula=formula_additive, data=df).fit()

# Step 4: Print summary of the model
print(log_reg_additive.summary())

# Step 5: Define and fit a synergistic model with an interaction term (e.g., Age * Female interaction)
formula_synergistic = 'Feel_connected_binary ~ Age * Female + C(Education_level)'
log_reg_synergistic = smf.logit(formula=formula_synergistic, data=df).fit()

# Print summary of the synergistic model
print(log_reg_synergistic.summary())


C:\Users\eznam\AppData\Local\Temp\ipykernel_15820\1242305144.py:6: DtypeWarning: Columns (408,1001,1002,1006,1007,1008,1080,1113,1115,1116,1117,1118,1119,1120,1121,1124,1125,1126,1127,1128,1213,1214,1215,1216,1217,1218,1342,1343,1344,1345,1346,1347,1348,1349,1390,1391,1393,1463,1549,1552,1555,1558,1561) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(url)


KeyError: 'Feel_connected_yn'

### **Question 4**


In [3]:
import pandas as pd
import statsmodels.formula.api as smf

url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
# fail https://github.com/KeithGalli/pandas/blob/master/pokemon_data.csv
pokeaman = pd.read_csv(url) 
pokeaman

model1_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation) + Q("Sp. Def"):C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") * C(Generation)', data=pokeaman)

model2_fit = model2_spec.fit()
model2_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.176
Model:                            OLS   Adj. R-squared:                  0.164
Method:                 Least Squares   F-statistic:                     15.27
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           3.50e-27
Time:                        18:48:49   Log-Likelihood:                -3649.4
No. Observations:                 800   AIC:                             7323.
Df Residuals:                     788   BIC:                             7379.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
===================================================================================================
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                          26.8971      5.246      5.127      0.000      16.599      37.195
C(Generation)[T.2]                 20.0449      7.821      2.563      0.011       4.692      35.398
C(Generation)[T.3]                 21.3662      6.998      3.053      0.002       7.629      35.103
C(Generation)[T.4]                 31.9575      8.235      3.881      0.000      15.793      48.122
C(Generation)[T.5]                  9.4926      7.883      1.204      0.229      -5.982      24.968
C(Generation)[T.6]                 22.2693      8.709      2.557      0.011       5.173      39.366
Q("Sp. Def")                        0.5634      0.071      7.906      0.000       0.423       0.703
Q("Sp. Def"):C(Generation)[T.2]    -0.2350      0.101     -2.316      0.021      -0.434      -0.036
Q("Sp. Def"):C(Generation)[T.3]    -0.3067      0.093     -3.300      0.001      -0.489      -0.124
Q("Sp. Def"):C(Generation)[T.4]    -0.3790      0.105     -3.600      0.000      -0.586      -0.172
Q("Sp. Def"):C(Generation)[T.5]    -0.0484      0.108     -0.447      0.655      -0.261       0.164
Q("Sp. Def"):C(Generation)[T.6]    -0.3083      0.112     -2.756      0.006      -0.528      -0.089
==============================================================================
Omnibus:                      337.229   Durbin-Watson:                   1.505
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2871.522
Skew:                           1.684   Prob(JB):                         0.00
Kurtosis:                      11.649   Cond. No.                     1.40e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

The apparent contradiction is between:

1. **R-squared (0.176)** indicates that 17.6% of variance in Pokemon HP is explained by the model. This might seem low.

2. **Statistically significant coefficients**: Many coefficients, including Generation indicators and their interactions with Sp. Def, have very small p-values (e.g., c(Generation)[2] has p=0.011) and relatively large coefficients (like 20.0449 for c(Generation)[2]).

These metrics address different aspects:
- R-squared evaluates the model's overall explanatory power
- Coefficients and p-values measure the strength and reliability of individual predictor effects

While most Generation indicators and some of their interactions with Sp. Def significantly impact Pokemon HP (as shown by small p-values), the overall model explains a modest portion of HP variance. This suggests other important factors affecting Pokemon HP aren't captured in this model.

### Question 5

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split

fifty_fifty_split_size = int(pokeaman.shape[0]*0.5)

# Replace "NaN" (in the "Type 2" column with "None")
pokeaman.fillna('None', inplace=True)

np.random.seed(130)
pokeaman_train,pokeaman_test = \
  train_test_split(pokeaman, train_size=fifty_fifty_split_size)
pokeaman_train

#                     Name   Type 1   Type 2  HP  Attack  Defense  \
370  338                  Solrock     Rock  Psychic  70      95       85   
6      6                Charizard     Fire   Flying  78      84       78   
242  224                Octillery    Water     None  75     105       75   
661  600                    Klang    Steel     None  60      80       95   
288  265                  Wurmple      Bug     None  45      45       35   
..   ...                      ...      ...      ...  ..     ...      ...   
522  471                  Glaceon      Ice     None  65      60      110   
243  225                 Delibird      Ice   Flying  45      55       45   
797  720      HoopaHoopa Confined  Psychic    Ghost  80     110       60   
117  109                  Koffing   Poison     None  40      65       95   
409  373  SalamenceMega Salamence   Dragon   Flying  95     145      130   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
370       55       65     70           3      False  
6        109       85    100           1      False  
242      105       75     45           2      False  
661       70       85     50           5      False  
288       20       30     20           3      False  
..       ...      ...    ...         ...        ...  
522      130       95     65           4      False  
243       65       45     75           2      False  
797      150      130     70           6       True  
117       60       45     35           1      False  
409      120       90    120           3      False  

[400 rows x 12 columns]

1. **Train-Test Split**: 
  - Pokemon dataset is evenly split into training and test sets
  - Missing values in "Type 2" are replaced with "None"
  - Most Pokemon aren't legendary
  - Some sort of balance/correlation between attack and defense, as well as between the Special Attack and Special Defensee

---

In [6]:
model_spec3 = smf.ols(formula='HP ~ Attack + Defense', 
                      data=pokeaman_train)
model3_fit = model_spec3.fit()
model3_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.148
Model:                            OLS   Adj. R-squared:                  0.143
Method:                 Least Squares   F-statistic:                     34.40
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           1.66e-14
Time:                        18:56:35   Log-Likelihood:                -1832.6
No. Observations:                 400   AIC:                             3671.
Df Residuals:                     397   BIC:                             3683.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     42.5882      3.580     11.897      0.000      35.551      49.626
Attack         0.2472      0.041      6.051      0.000       0.167       0.327
Defense        0.1001      0.045      2.201      0.028       0.011       0.190
==============================================================================
Omnibus:                      284.299   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5870.841
Skew:                           2.720   Prob(JB):                         0.00
Kurtosis:                      20.963   Cond. No.                         343.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

2. **Basic Model Performance**:
  - Simple model using Attack and Defense to predict HP achieves R² = 0.148
  - Both predictors are statistically significant (p < 0.05)
  - Model explains only 14.8% of HP variance

  ---

In [7]:
yhat_model3 = model3_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model3_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model3)[0,1]**2)

'In sample' R-squared:     0.14771558304519894
'Out of sample' R-squared: 0.21208501873920738


3. **Model Validation**:
  - In-sample R² = 0.148 matches model fit
  - Out-of-sample R² = 0.212 shows slightly better performance on test data
  - Reasonable generalization despite low overall predictive power

  ---

In [8]:
model4_linear_form = 'HP ~ Attack * Defense * Speed * Legendary'
model4_linear_form += ' * Q("Sp. Def") * Q("Sp. Atk")'
# DO NOT try adding '* C(Generation) * C(Q("Type 1")) * C(Q("Type 2"))'
# That's 6*18*19 = 6*18*19 possible interaction combinations...
# ...a huge number that will blow up your computer

model4_spec = smf.ols(formula=model4_linear_form, data=pokeaman_train)
model4_fit = model4_spec.fit()
model4_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.467
Model:                            OLS   Adj. R-squared:                  0.369
Method:                 Least Squares   F-statistic:                     4.764
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           4.23e-21
Time:                        18:57:27   Log-Likelihood:                -1738.6
No. Observations:                 400   AIC:                             3603.
Df Residuals:                     337   BIC:                             3855.
Df Model:                          62                                         
Covariance Type:            nonrobust                                         
====================================================================================================================================
                                                                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------------------
Intercept                                                          521.5714    130.273      4.004      0.000     265.322     777.821
Legendary[T.True]                                                   -6.1179      2.846     -2.150      0.032     -11.716      -0.520
Attack                                                              -8.1938      2.329     -3.518      0.000     -12.775      -3.612
Attack:Legendary[T.True]                                         -1224.9589    545.105     -2.247      0.025   -2297.197    -152.721
Defense                                                             -6.1989      2.174     -2.851      0.005     -10.475      -1.923
Defense:Legendary[T.True]                                         -102.4028     96.565     -1.060      0.290    -292.350      87.544
Attack:Defense                                                       0.0985      0.033      2.982      0.003       0.034       0.164
Attack:Defense:Legendary[T.True]                                    14.6361      6.267      2.336      0.020       2.310      26.963
Speed                                                               -7.2261      2.178     -3.318      0.001     -11.511      -2.942
Speed:Legendary[T.True]                                            704.8786    337.855      2.086      0.038      40.308    1369.449
Attack:Speed                                                         0.1264      0.038      3.351      0.001       0.052       0.201
Attack:Speed:Legendary[T.True]                                       5.8648      2.692      2.179      0.030       0.570      11.160
Defense:Speed                                                        0.1026      0.039      2.634      0.009       0.026       0.179
Defense:Speed:Legendary[T.True]                                     -6.9266      3.465     -1.999      0.046     -13.742      -0.111
Attack:Defense:Speed                                                -0.0016      0.001     -2.837      0.005      -0.003      -0.001
Attack:Defense:Speed:Legendary[T.True]                              -0.0743      0.030     -2.477      0.014      -0.133      -0.015
Q("Sp. Def")                                                        -5.3982      1.938     -2.785      0.006      -9.211      -1.586
Legendary[T.True]:Q("Sp. Def")                                    -282.2492    126.835     -2.225      0.027    -531.738     -32.760
Attack:Q("Sp. Def")                                                  0.1094      0.034      3.233      0.001       0.043       0.176
Attack:Legendary[T.True]:Q("Sp. Def")                               12.6502      5.851      2.162      0.031       1.141      24.160
Defense:Q("Sp. Def")    

4. **Complex Interaction Model**:
  - Includes interactions between Attack, Defense, Speed, Legendary status, Sp. Atk, and Sp. Def
  - Higher in-sample R² = 0.467 but much lower adjusted R² = 0.369
  - Many interaction terms are statistically significant (p < 0.05)
  - Warning of strong multicollinearity (condition number = 1.2e+16)
---

In [9]:
yhat_model4 = model4_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model4_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model4)[0,1]**2)

'In sample' R-squared:     0.4670944211580673
'Out of sample' R-squared: 0.0024853451035464894


5. **Overfitting Assessment**:
  - In-sample R² = 0.467 shows decent training performance
  - Out-of-sample R² = 0.002 reveals severe overfitting
  - Complex model fails to generalize despite higher training R²

---

### Question 6

✅

### **Question 7**

The development from Model 3 to Model 7 follows a systematic refinement process:

**Initial Models**:
- Model 3: Basic linear regression with Attack and Defense
- Model 4: Overly complex with many interactions, showing severe multicollinearity (condition number ~12e+16)

**Refinement Process**:
1. Model 5:
  - Added main effects (Attack, Defense, Speed, Legendary, Sp. Def, Sp. Atk)
  - Included categorical variables (Generation, Type 1, Type 2)
  - Avoided interactions to prevent multicollinearity

2. Model 6:
  - Retained significant predictors from Model 5
  - Added specific categorical indicators (Normal/Water types, Generations 2/5)
  - Removed less significant variables based on hypothesis testing

3. Model 7:
  - Introduced targeted interactions between continuous variables
  - Maintained significant categorical indicators
  - Applied centering and scaling to continuous predictors
  - Dramatically reduced multicollinearity (condition number: 15.4)

**Key Improvements**:
- Progressive refinement based on statistical evidence
- Balance between model complexity and predictive power
- Careful management of multicollinearity through variable selection and scaling
- Focus on generalizable relationships rather than overfitting

The final model (Model 7) achieves better out-of-sample performance while maintaining interpretability and avoiding the severe multicollinearity issues present in earlier iterations.

### Question 8
This was all ChatGPT

In [ ]:
import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
from sklearn.model_selection import train_test_split
import plotly.express as px
import plotly.graph_objects as go

# Number of repetitions
reps = 100

# Arrays to store R-squared values for each repetition
in_sample_Rsquared = np.zeros(reps)
out_of_sample_Rsquared = np.zeros(reps)

# Define the linear form based on model3 specification (you can change it to other models if needed)
linear_form = 'HP ~ Attack + Defense + Speed + Q("Sp. Def") + Q("Sp. Atk")'

# Loop over the number of repetitions
for i in range(reps):
    # Random 50-50 train-test split
    pokeaman_train, pokeaman_test = train_test_split(pokeaman, test_size=0.5)

    # Fit the model on the training set
    model_spec = smf.ols(formula=linear_form, data=pokeaman_train)
    model_fit = model_spec.fit()

    # Calculate in-sample R-squared
    in_sample_Rsquared[i] = model_fit.rsquared

    # Calculate out-of-sample R-squared
    y_test = pokeaman_test['HP']
    y_pred_test = model_fit.predict(pokeaman_test)
    out_of_sample_Rsquared[i] = np.corrcoef(y_test, y_pred_test)[0, 1] ** 2

# Create a DataFrame with the collected R-squared values
df = pd.DataFrame({
    "In Sample R-squared": in_sample_Rsquared,
    "Out of Sample R-squared": out_of_sample_Rsquared
})

# Plot the results
fig = px.scatter(
    df, 
    x="In Sample R-squared", 
    y="Out of Sample R-squared",
    labels={"In Sample R-squared": "In-Sample R²", "Out of Sample R-squared": "Out-of-Sample R²"},
    title="Comparison of In-Sample vs Out-of-Sample R-squared"
)

# Add a reference line y = x for comparison
fig.add_trace(
    go.Scatter(
        x=[0, 1], 
        y=[0, 1], 
        mode='lines', 
        line=dict(color='red', dash='dash'),
        name='y = x (Reference Line)'
    )
)

# Display the plot
# fig.show()
fig.show(renderer="png")

### **Question 9**

✅


---

### Summary

Here's a summary of the main points:

Definitions of Key Terms:

- Simple Linear Regression, Multiple Linear Regression, Continuous Variable, Indicator Variable, and Linear Forms.
- Simple and Multiple Linear Regressions are explained in terms of modeling relationships between dependent and independent variables.
- Continuous Variables represent measurable values, while Indicator Variables are binary, often used to include categorical data in regression models.
- Linear Forms are expressions used in modeling linear relationships.
- Regression Analysis Guidance:

- Instructions for using statistical models like statsmodels and pandas to perform regression, including creating binary outcomes, applying linear forms, and using plotly for visualization.
- Example code is provided to demonstrate how to create and fit linear and logistic regression models, interpret coefficients, and generate visualizations.
- Understanding R-squared and P-values:

**Explanation of R-squared (variance explained by the model) and p-values (significance of individual predictors).**
- Discussion on how high significance (low p-values) can coexist with low R-squared, as each measures different aspects of model fit.
- Addressing Multicollinearity:

**Instructions on calculating the condition number to check for multicollinearity, as excessive multicollinearity affects model stability and generalizability.**
- Emphasis on centering and scaling predictors to mitigate multicollinearity.
- Iterative Model Refinement:

- Guidelines for building and refining models through incremental complexity, aiming for balance between capturing significant patterns and avoiding overfitting.
